sounddevice.py

```diff
@@ -2289,14 +2289,41 @@ def reset(self):
         vars(self).clear()
         self.__init__()
 
+
 if not hasattr(_ffi, 'I_AM_FAKE'):
     # This object shadows the 'default' class, except when building the docs.
     default = default()
 
 
 class PortAudioError(Exception):
-    """This exception will be raised on PortAudio errors."""
+    """This exception will be raised on PortAudio errors.
+    
+    Attributes
+    ----------
+    args
+        A variable length tuple containing the following parameters when
+        available:
+
+        error_message
+            The user error message.
+        pa_error_code
+            The PortAudio `PaErrorCode` value.
+        hosterror_info
+            A 3-tuple containing the host API index, host error code, and
+            the host error message or a zero length string if no message is
+            available.
 
+    """
+    def __str__(self):
+        errormsg = self.args[0] if len(self.args) else ''
+        if len(self.args) > 1:
+            errormsg = "{} [PaErrorCode {}]".format(errormsg, self.args[1])
+        if len(self.args) > 2:
+            host_api, hosterror_code, hosterror_text = self.args[2]
+            hostname = query_hostapis(host_api)['name']
+            errormsg = "{}: '{}' [{} error {}]".format(errormsg, hosterror_text,
```

Comment:
I guess both are OK and both have their flaws.

I wasn't aware of the `u` prefix, but my policy regarding Python 2 is: I want it to work (for now, until support is dropped), but I don't care if it is ugly.

For me, using `{1!r}` is habit.
The reason for that is that the quotes automatically change if the strings contains single quotes.
The downside is that "special" characters like `\n` are shown with their escape character ... which is good in many cases, but here it's probably not ...

So in the end it looks like your way is better here!